File: dotnet/DesignPatterns/BehavioralDesignPatterns/StateDesignPattern/StateDesignPattern.Example1/Order.cs
namespace StateDesignPattern.Example1;

/// <summary>
/// Context - maintains a reference to the current state and delegates state-specific behavior.
/// </summary>
public class Order
{
    private IOrderState _currentState;
    public string OrderId { get; }
    public string ProductName { get; }

    public Order(string orderId, string productName)
    {
        OrderId = orderId;
        ProductName = productName;
        _currentState = new PendingState();
    }

    public void SetState(IOrderState state)
    {
        _currentState = state;
    }

    public string GetCurrentStateName()
    {
        return _currentState.GetStateName();
    }

    public void Process()
    {
        _currentState.ProcessOrder(this);
    }

    public void Ship()
    {
        _currentState.ShipOrder(this);
    }

    public void Deliver()
    {
        _currentState.DeliverOrder(this);
    }

    public void Cancel()
    {
        _currentState.CancelOrder(this);
    }

    public void PrintStatus()
    {
        Console.WriteLine($"Order [{OrderId}] - Product: {ProductName} - Status: {GetCurrentStateName()}");
    }
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/StateDesignPattern/StateDesignPattern.Example1/src/Order.ts
import { IOrderState } from './IOrderState';
import { PendingState } from './PendingState';

export class Order {
  private _currentState: IOrderState;
  readonly orderId: string;
  readonly productName: string;

  constructor(orderId: string, productName: string) {
    this.orderId = orderId;
    this.productName = productName;
    this._currentState = new PendingState();
  }

  setState(state: IOrderState): void {
    this._currentState = state;
  }

  getCurrentStateName(): string {
    return this._currentState.getStateName();
  }

  process(): void {
    this._currentState.processOrder(this);
  }

  ship(): void {
    this._currentState.shipOrder(this);
  }

  deliver(): void {
    this._currentState.deliverOrder(this);
  }

  cancel(): void {
    this._currentState.cancelOrder(this);
  }

  printStatus(): void {
    console.log(`Order [${this.orderId}] - Product: ${this.productName} - Status: ${this.getCurrentStateName()}`);
  }
}
